### PR TITLE
feature(relationships): it's now possible to directly query a set of relationships

### LINKED ISSUE #7299: FR: a function to get relationships (eg elgg_get_relationships)
I would like to get a function to return relationships. 

Examples of use cases:
- get all friend relationships (so NOT friends of user X, but all relationships)
- all group memberships
- all relationships of user X (this will include group memberships and friends)

possible filters are
- guid_one
- guid_two
- relationship
- time limitations based on time_created of the relationship

result will be an array of ElggRelationship's

### PR DESCRIPTION
**Pull request by juho-jaakkola**
Earlier you needed to first get entities from a relationship and then process relationships one by one for each entity. Now it is possible to query the relationships directly.

Fixes #7299

### CODE REVIEW

File: engine/classes/Elgg/Database/RelationshipsTable.php [hunk 1]
```diff
@@ -249,6 +249,64 @@ public function removeAll($guid, $relationship = "", $inverse_relationship = fal
...
+
+		$options = array_merge($defaults, $options);
+
+		$wheres = array();
+
+		if (isset($options['relationship'])) {
+			$names = array_map(function($value) {
+				$string = $this->db->sanitizeString($value);
+				return "'{$string}'";
+			}, $options['relationship']);
```
Comment (hypeJunction) [inline engine/classes/Elgg/Database/RelationshipsTable.php]: Is it normalized to array somewhere?

Comment (juho-jaakkola) [inline engine/classes/Elgg/Database/RelationshipsTable.php]: Nope, but probably should.

**Comment by juho-jaakkola** [comment]
On a side note: ugh, we really need a database abstraction layer.

File: engine/classes/Elgg/Database/RelationshipsTable.php (hunk 1 above)
Comment (juho-jaakkola) [inline engine/classes/Elgg/Database/RelationshipsTable.php]: Done

**Comment by hypeJunction** [comment]
LGTM

**Comment by juho-jaakkola** [comment]
This needs more filters. I wonder what would be a good compromise between flexibility and complexity.

Allowing `wheres` and `joins` in the options would pretty much cover all possible needs, but they make the function much more complicated.

**Comment by mrclay** [comment]
Abilities to offset or remove the limit seem like natural needs, but that also means we should order by something, which means... Do we have a foreseen use case that would help?

You could try using DBAL's QueryBuilder.

**Comment by mrclay** [comment]
Sorry, plenty of use cases already listed in #7299.

**Comment by mrclay** [comment]
FYI I'm reworking this using QueryBuilder.

**Comment by juho-jaakkola** [comment]
Nice! Looking forward to see how it will turn out.

**Comment by mrclay** [comment]
Replaced by #9456

**Pull request closed without merging**